twirllib/src/mill/twirllib/TwirlWorker.scala

```diff
@@ -0,0 +1,109 @@
+package mill
+package twirllib
+
+import java.io.File
+import java.net.URLClassLoader
+
+import ammonite.ops.{Path, ls}
+import mill.eval.PathRef
+import mill.scalalib.CompilationResult
+
+import scala.io.Codec
+import scala.util.Properties
+
+class TwirlWorker {
+
+  private var twirlInstanceCache = Option.empty[(Long, TwirlWorkerApi)]
+
+  private def twirl(twirlClasspath: Agg[Path]) = {
+    val classloaderSig = twirlClasspath.map(p => p.toString().hashCode + p.mtime.toMillis).sum
+    twirlInstanceCache match {
+      case Some((sig, instance)) if sig == classloaderSig => instance
+      case _ =>
+        val cl = new URLClassLoader(twirlClasspath.map(_.toIO.toURI.toURL).toArray)
+        val twirlCompilerClass = cl.loadClass("play.twirl.compiler.TwirlCompiler")
+        val compileMethod = twirlCompilerClass.getMethod("compile",
+          classOf[java.io.File],
+          classOf[java.io.File],
+          classOf[java.io.File],
+          classOf[java.lang.String],
+          cl.loadClass("scala.collection.Seq"),
+          cl.loadClass("scala.collection.Seq"),
+          cl.loadClass("scala.io.Codec"),
+          classOf[Boolean])
+
+        val defaultAdditionalImportsMethod = twirlCompilerClass.getMethod("compile$default$5")
+        val defaultConstructorAnnotationsMethod = twirlCompilerClass.getMethod("compile$default$6")
+        val defaultCodecMethod = twirlCompilerClass.getMethod("compile$default$7")
+        val defaultFlagMethod = twirlCompilerClass.getMethod("compile$default$8")
+
+        val instance = new TwirlWorkerApi {
+          override def compileTwirl(source: File,
+                                    sourceDirectory: File,
+                                    generatedDirectory: File,
+                                    formatterType: String,
+                                    additionalImports: Seq[String] = Nil,
+                                    constructorAnnotations: Seq[String] = Nil,
+                                    codec: Codec = Codec(Properties.sourceEncoding),
+                                    inclusiveDot: Boolean = false) {
+            val o = compileMethod.invoke(null, source,
+              sourceDirectory,
+              generatedDirectory,
+              formatterType,
+              defaultAdditionalImportsMethod.invoke(additionalImports),
```

Comment:
How does it work, can you explain a little bit? Why do we invoke method with `additionalImports` parameter?

Reply:
@rockjam Good catch! It's not working :disappointed: 
`defaultAdditionalImportsMethod` is a method that does not take any argument and the call is wrong (ie. the first argument must be the object):

```scala
defaultAdditionalImportsMethod.invoke(null, additionalImports)
```

If I replace `defaultAdditionalImportsMethod.invoke(null)` with `additionalImports` then the reflection does not work anymore and the following exception is thrown:

```
java.lang.IllegalArgumentException: argument type mismatch
```

Back to square one I guess :cry: 
So it's working but we cannot define/override `additionalImports`, `constructorAnnotations`, `codec`, or `inclusiveDot`.

Reply:
Interestingly, `additionalImports` is an instance of `scala.collection.immutable.$colon$colon` not `scala.collection.immutable.Seq[String]`...

Reply:
We can keep the defaults for these arguments for now, and merge this PR like that. And we can address modifying of these arguments in subsequent PR. How about that?

Reply:
Ok, I will revert my changes to invoke the method without argument

Reply:
@rockjam Done! I've created (private) functions in `TwirlModule` and add a comment so when we know how to override these arguments, everything will be ready.